src/main/php/lang/ast/CompilingClassloader.class.php

```diff
@@ -7,18 +7,27 @@
 use lang\ElementNotFoundException;
 use lang\IClassLoader;
 use lang\XPClass;
+use lang\reflect\Package;
 
 class CompilingClassLoader implements IClassLoader {
   const EXTENSION = '.php';
 
   private static $instance= [];
+  public static $syntax= [];
   private $version;
 
   /** Creates a new instances with a given PHP runtime */
   private function __construct($emit) {
     $this->version= $emit->getSimpleName();
-    Compiled::$emit[$this->version]= $emit;
 
+    $emitter= $emit->newInstance();
+    $language= Language::named('PHP');
+    foreach (self::$syntax as $syntax) {
+      $syntax->setup($language, $emitter);
+    }
```

Comment:
The syntaxes are really extensions to PHP, both on input language and emitter result. They are not generic - they wouldn't work if we had a `Language::named('JavaScript')` or an emitter which would output GoLang.

Reply:
```php
$emitter->transform('using', function($node) {
  $statement= new Symbol();
  $statement->std= null;
  $cleanup= [];
  foreach ($node->value->arguments as $i => $expression) {
    switch ($expression->kind) {
      case 'variable':
        $variable= $expression;
        yield $expression;
        break;

      case 'assignment':
        $variable= $expression->value->variable;
        yield $expression;
        break;

      default:
        $variable= new Node($expression->symbol, 'variable', 'U'.$i);
        yield new Node($statement, 'assignment', new Assignment($variable, '=', $expression));
        break;
    }

    $close= new Node($statement, 'invoke', new InvokeExpression(
      new Node($node->symbol, 'instance', new InstanceExpression($variable, new Node($node->symbol, 'name', 'close'))),
      []
    ));

    $dispose= new Node($statement, 'invoke', new InvokeExpression(
      new Node($node->symbol, 'instance', new InstanceExpression($variable, new Node($node->symbol, 'name', '__dispose'))),
      []
    ));

    $cleanup[]= new Node($node->symbol, 'if', new IfStatement(
      new Node($node->symbol, 'instanceof', new InstanceOfExpression($variable, '\lang\Closeable')),
      [$close],
      new Node($node->symbol, 'if', new IfStatement(
        new Node($node->symbol, 'instanceof', new InstanceOfExpression($variable, '\IDisposable')),
        [$dispose],
        null
      ))
    ));
    $cleanup[]= new Node($statement, 'invoke', new InvokeExpression(new Node($node->symbol, 'name', 'unset'), [$variable]));
  }

  yield new Node($node->symbol, 'try', new TryStatement($node->value->body, [], $cleanup));
});
```

Reply:
Has been refactored now to do without the `Node`s everywhere, see https://github.com/xp-framework/ast/pull/2

Reply:
This now becomes:

```php
$emitter->transform('using', function($node) {
  $cleanup= [];
  foreach ($node->arguments as $i => $expression) {
    switch ($expression->kind) {
      case 'variable': $variable= $expression; yield $expression; break;
      case 'assignment': $variable= $expression->variable; yield $expression; break;
      default: $variable= new Variable('U'.$i); yield new Assignment($variable, '=', $expression); break;
    }

    $cleanup[]= new IfStatement(new InstanceOfExpression($variable, '\lang\Closeable'),
      [new InvokeExpression(new InstanceExpression($variable, new Literal('close')), [])],
      [new IfStatement(new InstanceOfExpression($variable, '\IDisposable'),
        [new InvokeExpression(new InstanceExpression($variable, new Literal('__dispose')), [])]
      )]
    );
    $cleanup[]= new InvokeExpression(new Literal('unset'), [$variable]);
  }

  yield new TryStatement($node->body, null, $cleanup);
});
```